### PR TITLE
Async Invoker API

### PR DESCRIPTION
**Pull request by lorban**
TC API changes that introduce an async invoker API

### CODE REVIEW

File: entity-client-api/src/main/java/org/terracotta/entity/AsyncInvocationBuilder.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+ */
+public interface AsyncInvocationBuilder<M extends EntityMessage, R extends EntityResponse> {
+
+  /**
+   * <p>Sets whether or not the invocation should be replicated to any passive servers in this stripe.</p>
+   *
+   * <p>If this method call is omitted from builder, the default behavior is to replicate the message.</p>
+   *
+   * @param requiresReplication True if the message should be replicated, false otherwise
+   * @return Itself
```
Comment (chrisdennis) [inline entity-client-api/src/main/java/org/terracotta/entity/AsyncInvocationBuilder.java]: probably good to use the word `this` here?

File: entity-client-api/src/main/java/org/terracotta/entity/AsyncInvocationBuilder.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+   */
+  AsyncInvocationBuilder<M, R> message(M message);
+
+  /**
+   * <p>Block the {@link #invoke(InvocationCallback)} call until the message was enqueued or until timeout.</p>
+   * @param time the timeout value.
+   * @param unit the timeout unit.
+   * @return Itself
+   */
+  AsyncInvocationBuilder<M, R> blockEnqueuing(long time, TimeUnit unit);
```
Comment (chrisdennis) [inline entity-client-api/src/main/java/org/terracotta/entity/AsyncInvocationBuilder.java]: Should we document the default for this somewhere?

File: entity-client-api/src/main/java/org/terracotta/entity/InvocationCallback.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+ * <ul>
+ *  <li><b>SENT</b> means that a message has become "in-flight", meaning that it will be re-sent in the event of a
+ *  server restart or fail-over to passive. Note that this message is not generally useful to applications</li>
+ *  <li><b>RECEIVED</b> if requested, will only return from invoke() once the global order of this message (relative to
+ *  others depending on RECEIVED) has been determined for the entire stripe. This means that every server in the stripe
+ *  has received the message, and agreed upon the order. Note that RECEIVED is the only optional ack (the others always
+ *  go over the wire, but only optionally influence client-side waiting). This should be requested in cases where
+ *  knowing the global order of messages (preserved for restart or fail-over) is important</li>
+ *  <li><b>COMPLETED</b> is observed by the client when the active has finished running the invoke and returned the
+ *  result. Calling get() on the returned InvokeFuture will, by default, wait until RETIRED but the InvocationBuilder
```
Comment (chrisdennis) [inline entity-client-api/src/main/java/org/terracotta/entity/InvocationCallback.java]: This sentence doesn't make sense in an async context right?

File: entity-client-api/src/main/java/org/terracotta/entity/InvocationCallback.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+ *  can be configured to make the get() return only once COMPLETED has been received. Note that this only relates to
+ *  what has happened on the active: it is possible for the client to see the COMPLETED ack before any passives have
+ *  seen the message</li>
+ *  <li><b>RETIRED</b> is the final acknowledgement, after which time the message will not be re-sent in the event of
+ *  restart or fail-over. It represents the point where the message has completed on all servers in the stripe and also
+ *  any messages it created to defer its retirement (see {@link IEntityMessenger}) have also completed on all servers in
+ *  the stripe</li>
+ *  <li><b>RESULT</b> is the decoded response returned by the server</li>
```
Comment (chrisdennis) [inline entity-client-api/src/main/java/org/terracotta/entity/InvocationCallback.java]: `responses`in the plural, right?

File: entity-client-api/src/main/java/org/terracotta/entity/InvocationCallback.java [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+ *  <li><b>FAILURE</b> is the error that occurred along the way</li>
```
Comment (chrisdennis) [inline entity-client-api/src/main/java/org/terracotta/entity/InvocationCallback.java]: `error` or `errors`?

File: entity-client-api/src/main/java/org/terracotta/entity/InvocationCallback.java [hunk 6]
```diff
@@ -0,0 +1,45 @@
+package org.terracotta.entity;
+
+/**
+ * <p>The acknowledgement configurations can be complex so a high-level description of them and their meanings is
+ * provided here, with greater detail in the relevant methods, below:</p>
```
Comment (chrisdennis) [inline entity-client-api/src/main/java/org/terracotta/entity/InvocationCallback.java]: There is no greater detail in the methods atm.

File: entity-client-api/src/main/java/org/terracotta/entity/AsyncInvocationBuilder.java (hunk 1 above)
Comment (lorban) [reply entity-client-api/src/main/java/org/terracotta/entity/AsyncInvocationBuilder.java]: `InvocationBuilder` documents that it returns `Itself` hence I followed that convention. But you're right, `this` is more correct so I changed both this class as well as `InvocationBuilder`.

File: entity-client-api/src/main/java/org/terracotta/entity/AsyncInvocationBuilder.java (hunk 2 above)
Comment (lorban) [reply entity-client-api/src/main/java/org/terracotta/entity/AsyncInvocationBuilder.java]: Will do.

File: entity-client-api/src/main/java/org/terracotta/entity/InvocationCallback.java (hunk 3 above)
Comment (lorban) [reply entity-client-api/src/main/java/org/terracotta/entity/InvocationCallback.java]: Will fix.

**Pull request merged**